VMSS Flex support for MachinePools

mboersma opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Implements VMSS Flex mode support for MachinePools.

~~There is still work to do here, I'm just getting a WIP branch out to help with collaboration. For example:~~

- ~~Flex mode still on by default~~ Fixed
- ~~MP ReadyReplicas not always syncing with actual VM count~~ Fixed
- ~~faultDomainCount still hard-coded at 3 because cache code isn't working~~ Fixed

**Which issue(s) this PR fixes**:

Fixes #999
Fixes #2987

**Special notes for your reviewer**:

This work was started and mostly completed by @devigned and @jackfrancis. Thanks team!

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
VMSS Flex support for MachinePools
```

CecileRobertMichon commented on Tiltfile:
```diff
@@ -358,8 +358,10 @@ def deploy_worker_templates(template, substitutions):
-        flavor_cmd += "; " + helm_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig install --repo https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/helm/repo cloud-provider-azure --generate-name --set infra.clusterName=${CLUSTER_NAME}"
+    if "external-cloud-provider" in flavor_name or "flex" in flavor_name:
+        flavor_cmd += "; " + helm_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig install --repo https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/helm/repo cloud-provider-azure --generate-name --set infra.clusterName=${CLUSTER_NAME}" + \
+                      " --set cloudControllerManager.imageRepository=selfhostedk8s.azurecr.io --set cloudControllerManager.imageTag=d8d65f1" + \
```
Can we add a TODO code comment to remove this once the release of cloud-provider we're waiting for gets released?

CecileRobertMichon commented on azure/converters/vmss.go:
```diff
@@ -61,6 +61,38 @@ func SDKToVMSS(sdkvmss compute.VirtualMachineScaleSet, sdkinstances []compute.Vi
 }
 
+// SDKVMToVMSSVM converts an Azure SDK VM to a VMSS VM.
+func SDKVMToVMSSVM(sdkInstance compute.VirtualMachine) *azure.VMSSVM {
```
missing unit test for this?

CecileRobertMichon commented on azure/scope/machinepool.go:
```diff
@@ -19,8 +19,11 @@ package scope
+	"strings"
 	"time"
 
+	azure2 "github.com/Azure/go-autorest/autorest/azure"
```
do we have a better convention for naming this anywhere else in the repo already? azure2 isn't very explicit. If so this would be a good one to add to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/.golangci.yml#L59

mboersma replied:
See #2958

CecileRobertMichon commented on azure/scope/machinepool.go:
```diff
@@ -330,17 +334,14 @@ func (m *MachinePoolScope) applyAzureMachinePoolMachines(ctx context.Context) er
 	ampm := infrav1exp.AzureMachinePoolMachine{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      m.AzureMachinePool.Name + "-" + machine.InstanceID,
+			Name:      strings.ReplaceAll(parsed.ResourceName, "_", "-"),
```
could this break existing machines? just wondering if we're introducing a new way of naming machines that would change exisitng machines

CecileRobertMichon commented on azure/services/scalesets/scalesets.go:
```diff
@@ -271,10 +270,23 @@ func (s *Service) patchVMSSIfNeeded(ctx context.Context, infraVMSS *azure.VMSS)
+			break
+		}
+	}
+	var hasEnoughLatestModelOrNotMixedModel bool
```
Not sure what `hasEnoughLatestModelOrNotMixedModel ` means at first read. Consider adding a code comment to explain?

CecileRobertMichon commented on azure/services/scalesets/scalesets.go:
```diff
@@ -795,3 +810,10 @@ func getSecurityProfile(vmssSpec azure.ScaleSetSpec, sku resourceskus.SKU) (*com
 	return true, nil
 }
+
+func getOrchestrationMode(modeType infrav1.OrchestrationModeType) compute.OrchestrationMode {
```
this probably belongs better in the converters package

CecileRobertMichon replied:
also let's add a simple unit test even if it's trivial

CecileRobertMichon commented on azure/services/scalesetvms/scalesetvms.go:
```diff
@@ -122,24 +237,53 @@ func (s *Service) Delete(ctx context.Context) error {
+// Parameters takes the existing resource and returns the desired parameters of the resource.
+// If the resource does not exist, or we do not care about existing parameters to update the resource, existing should be nil.
+// If no update is needed on the resource, Parameters should return nil.
+func (vm *VMSSFlexVMGetter) Parameters(existing interface{}) (params interface{}, err error) {
```
what's going on with this no-op Parameters func here? Was this an attempt to implement the ResourceSpecGetter interface but without using it yet since we're not yet using the async package?

CecileRobertMichon commented on azure/services/virtualmachines/client.go:
```diff
@@ -31,11 +34,35 @@ import (
-	virtualmachines compute.VirtualMachinesClient
+type (
+	// Client wraps go-sdk.
+	Client interface {
```
Do we need this to interface to be exported? Is it used anywhere outside of this package?

mboersma replied:
Yes, it's used in [scalesetvms.go](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2813/files#diff-afa3aa43d3541f80b10af246e174440ced390fc557b6a907525183d4d9dd5106R51).

CecileRobertMichon commented on azure/types.go:
```diff
@@ -135,6 +137,13 @@ func (vmss VMSS) InstancesByProviderID() map[string]VMSSVM {
 // ProviderID returns the K8s provider ID for the VMSS instance.
 func (vm VMSSVM) ProviderID() string {
+	if vm.InstanceID == "" {
+		// build providerID a bit differently for VMSS Flex
```
What does "a bit differently" mean? this code is not very human friendly, a more explicit code comment with a concrete example could be helpful

mboersma commented:
/retitle VMSS Flex support for MachinePools

mboersma commented:
/retest

CecileRobertMichon commented on Tiltfile:
```diff
@@ -359,7 +359,9 @@ def deploy_worker_templates(template, substitutions):
     flavor_cmd += "; " + helm_cmd + " repo add projectcalico https://projectcalico.docs.tigera.io/charts; " + helm_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig install calico projectcalico/tigera-operator -f " + calico_values + " --namespace tigera-operator --create-namespace; kubectl --kubeconfig ./${CLUSTER_NAME}.kubeconfig apply -f ./templates/addons/calico/felix-override.yaml"
     if "external-cloud-provider" in flavor_name:
-        flavor_cmd += "; " + helm_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig install --repo https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/helm/repo cloud-provider-azure --generate-name --set infra.clusterName=${CLUSTER_NAME}"
+        # TODO: remove extra config once the azure-cloud-provider Helm chart release supports VMSS Flex.
```
Cloud-provider-azure released v1.26 last week so this should only be a matter of updating the Helm chart https://github.com/kubernetes-sigs/cloud-provider-azure/releases/tag/v1.26.0

cc @lzhecheng

mboersma replied:
That release did not update the Helm chart, so provisioning with Kubernetes v1.26.0 gives an error about `imageTag required` without this workaround.

CecileRobertMichon replied:
Right, I'm saying we should update the Helm chart (@lzhecheng @jackfrancis are usually the ones who maintain it but anyone can make a PR).

mboersma replied:
See kubernetes-sigs/cloud-provider-azure#2994

CecileRobertMichon commented on Tiltfile:
```diff
@@ -359,7 +359,9 @@ def deploy_worker_templates(template, substitutions):
-        flavor_cmd += "; " + helm_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig install --repo https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/helm/repo cloud-provider-azure --generate-name --set infra.clusterName=${CLUSTER_NAME}"
+        # TODO: remove extra config once the azure-cloud-provider Helm chart release supports VMSS Flex.
+        flavor_cmd += "; " + helm_cmd + " --kubeconfig ./${CLUSTER_NAME}.kubeconfig install --repo https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/helm/repo cloud-provider-azure --generate-name --set infra.clusterName=${CLUSTER_NAME}" + \
+                      " --set cloudControllerManager.imageTag=v1.26.0 --set cloudNodeManager.imageTag=v1.26.0"
```
this forces the version of k8s to 1.26 which might not be the case (in fact the tiltfile default for k8s version right now is 1.24 https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2813/files#diff-c2ee8653e1d6b85f0aadf87cd438a9250806c052877248442be4d434cbc52425R24). I think it might make sense to revert this change and then enforce "you must use k8s 1.26+ to enable vmss flex"

mboersma replied:
I added this restriction in documentation, but adding it to validation seems tricky. The natural place to do this is when validating an `AzureMachinePool` and seeing `orchestrationMode: Flexible`. 

But the Kubernetes `version` is on the owner `MachinePool`. I can find it by iterating the owners, but I'm not sure how to fetch the actual resource through the OwnerRef or if that would be appropriate in the validation code.

It would also be ideal if this validation could check for external cloud provider, but that faces similar challenges.

CecileRobertMichon replied:
Copy/Pasting since the answer is similar: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2965#discussion_r1061046246

Want to open a follow up issue to add validation via webhook?

mboersma replied:
See #2987.

CecileRobertMichon commented on azure/services/scalesetvms/scalesetvms.go:
```diff
@@ -143,3 +257,32 @@ func (s *Service) Delete(ctx context.Context) error {
+// Parameters takes the existing resource and returns the desired parameters of the resource.
+// If the resource does not exist, or we do not care about existing parameters to update the resource, existing should be nil.
+// If no update is needed on the resource, Parameters should return nil.
+func (vm *VMSSFlexVMGetter) Parameters(ctx context.Context, existing interface{}) (params interface{}, err error) {
```
can you add a comment here on why Parameters is a no-op? Assuming it's because #2720 hasn't been implemented yet

CecileRobertMichon commented on azure/services/virtualmachines/client.go:
```diff
@@ -169,3 +211,52 @@ func (ac *AzureClient) Result(ctx context.Context, future azureautorest.FutureAP
 }
+
+// GetResultIfDone fetches the result of a long-running operation future if it is done.
+func (ac *AzureClient) GetResultIfDone(ctx context.Context, future *infrav1.Future) (compute.VirtualMachine, error) {
```
we can probably clean this up a bit but let's leave that for #2720

CecileRobertMichon commented on templates/flavors/external-cloud-provider-machinepool/kustomization.yaml:
```diff
@@ -0,0 +1,10 @@
+namespace: default
+resources:
+  - ../base
```
If you start this from https://github.com/kubernetes-sigs/cluster-api-provider-azure/tree/main/templates/flavors/machinepool it would already have the machinepool and azureclusteridentity (then we can just patch it for flex and OOT)

CecileRobertMichon commented on templates/test/ci/prow-external-cloud-provider/kustomization.yaml:
```diff
@@ -2,10 +2,11 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 namespace: default
 resources:
-  - ../../../flavors/external-cloud-provider
+  - ../../../flavors/external-cloud-provider-machinepool
```
would it make sense to just have one template that has both md and mp to test OOT cloud provider? Seems like we're getting rid of md by doing this? The base prow template has both I believe

CecileRobertMichon replied:
Also, are we testing both Flex and Uniform?

mboersma replied:
I reverted this changed and instead created a new `prow-external-cloud-provider-machinepool` test template. It inherits from the `external-cloud-provider-machinepool` flavor, so they're both just testing VMSS Flex with Linux currently.

I also added a new e2e test spec for this. I would like to expand it to include both Flex and Uniform for Linux and Windows (4 node pools total), but it seems safer to start small.

CecileRobertMichon commented on templates/test/ci/cluster-template-prow-external-cloud-provider.yaml:
```diff
@@ -240,3 +243,17 @@ spec:
+apiVersion: addons.cluster.x-k8s.io/v1beta1
+kind: ClusterResourceSet
+metadata:
+  name: ${CLUSTER_NAME}-calico-windows
```
are there windows nodes in this template?

mboersma replied:
No, but this change should be reverted now.

CecileRobertMichon commented on test/e2e/cloud-provider-azure.go:
```diff
@@ -45,7 +45,13 @@ func InstallCalicoAndCloudProviderAzureHelmChart(ctx context.Context, input clus
+			fmt.Sprintf("infra.clusterName=%s", input.ConfigCluster.ClusterName),
+			fmt.Sprintf("cloudControllerManager.clusterCIDR=%s", strings.Join(cidrBlocks, `,`)),
+			// TODO: remove extra config once the azure-cloud-provider Helm chart release supports VMSS Flex.
+			"cloudControllerManager.imageTag=v1.26.0",
```
same comment as Tilt here, we should be able to fix this now that cloud provider 1.26 is out, but we shouldn't use v1.26 unless k8s version is also 1.26 as it may lead to incompatibility. This also means we can't test Flex unless we use k8s v1.26 in theory.

mboersma replied:
In theory, although in practice it seems to work fine with k8s 1.24 and 1.25. Previously we had it hard-coded to a pre-release SHA.

But I agree, we should limit this to k8s >= 1.26. I'll make that change.

CecileRobertMichon replied:
Could we also document this limitation and the requirement for OOT cloud-provider?

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/cc @devigned @lzhecheng @feiskyer @zmyzheng

mboersma commented:
/retest

mboersma commented:
/test pull-cluster-api-provider-azure-e2e-optional

devigned commented on azure/services/scalesets/scalesets.go:
```diff
@@ -517,6 +526,19 @@ func (s *Service) buildVMSSFromSpec(ctx context.Context, vmssSpec azure.ScaleSet
 		},
 	}
 
+	if orchestrationMode == compute.OrchestrationModeUniform {
```
total nit, but in this case, I'd probably lean toward a switch statement.

devigned commented on config/crd/bases/infrastructure.cluster.x-k8s.io_azuremachinepools.yaml:
```diff
@@ -1380,6 +1380,14 @@ spec:
                   NOTE: NodeDrainTimeout is different from `kubectl drain --timeout`'
                 type: string
+              orchestrationMode:
+                default: Uniform
```
I thought we discussed that we'd default to Flexible going forward and convert previous versions setting Uniform on them. I believe this line will default to Uniform if not specified.

mboersma replied:
When I discussed with @CecileRobertMichon, we thought it was safer to leave the default as `Uniform` and require users to opt in. Especially since VMSS Flex support requires out-of-tree cloud provider and Kubernetes >= 1.26.

devigned replied:
Makes sense. Thank you for the explanation.

devigned reviewed (changes requested):
Little nit, and quick clarification on defaulting of orchestration mode.

devigned reviewed (approved):
/lgtm

Glad to see this functionality nearing merge.

mboersma commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented on test/e2e/azure_test.go:
```diff
@@ -556,6 +556,71 @@ var _ = Describe("Workload cluster creation", func() {
+	Context("Creating a cluster that uses the external cloud provider and machinepools [OPTIONAL]", func() {
+		It("with 1 control plane node and 1 machinepool", func() {
+			By("using user-assigned identity")
+			clusterName = getClusterName(clusterNamePrefix, "oot")
```
```suggestion
			clusterName = getClusterName(clusterNamePrefix, "oot-flex")
```

so we can differentiate with the other oot cluster in logs

CecileRobertMichon commented:
lgtm however the flex test failed with a failure to provision the LB in CCM, @mboersma @zmyzheng PTAL

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2813/pull-cluster-api-provider-azure-e2e-optional/1610780765266644992/artifacts/clusters/capz-e2e-aclv4h-oot/kube-system/cloud-controller-manager-6555775b6d-4hwk6/cloud-controller-manager.log

mboersma commented:
I switched to an...errr...`switch` statement as suggested and renamed the test spec to "flex". Hopefully the e2e-optional test failure was a flake–it's been passing for me locally.

zmyzheng commented:
> lgtm however the flex test failed with a failure to provision the LB in CCM, @mboersma @zmyzheng PTAL
> 
> https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2813/pull-cluster-api-provider-azure-e2e-optional/1610780765266644992/artifacts/clusters/capz-e2e-aclv4h-oot/kube-system/cloud-controller-manager-6555775b6d-4hwk6/cloud-controller-manager.log

I took a look at the failed test and got a little confused: 
<img width="1683" alt="image" src="https://user-images.githubusercontent.com/24629574/210860434-2a1bf7ba-0ce8-4961-b287-e0ef44375136.png">

Is the test creating an external LB or Internal LB? The resource name is elb but the error shows it is create an internel LB somehow

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional
/retest

mboersma commented:
The LB test failed again on the Flex machinepool spec, I'll dig in to it.

mboersma commented:
@zmyzheng good eye--I hadn't noticed that. Maybe we're mislabeling something in the test spec? I'll look at it.

This is the specific line where the test failed: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/test/e2e/azure_lb.go#L213

mboersma commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional

jackfrancis commented on azure/services/scalesets/scalesets.go:
```diff
@@ -517,6 +526,19 @@ func (s *Service) buildVMSSFromSpec(ctx context.Context, vmssSpec azure.ScaleSet
+	case compute.OrchestrationModeFlexible:
+		vmss.VirtualMachineScaleSetProperties.VirtualMachineProfile.NetworkProfile.NetworkAPIVersion =
+			compute.NetworkAPIVersionTwoZeroTwoZeroHyphenMinusOneOneHyphenMinusZeroOne
+		vmss.VirtualMachineScaleSetProperties.PlatformFaultDomainCount = to.Int32Ptr(1)
```
So basically the VMSS Flex API exposes a new `PlatformFaultDomainCount` property that is a peer property to `Zones`, and we always want to ship a VMSS Flex config where those properties have the same value?

mboersma replied:
Yes, that's my understanding: `PlatformFaultDomainCount` is required for `OrchestrationMode: Flexible` and shouldn't be greater than the number of zones.

jackfrancis replied:
ACK

If we wanted to expose this to the user later will this current default enforcement be forward-compatible with such a change in the future? I think so.

jackfrancis commented on azure/services/scalesetvms/scalesetvms.go:
```diff
@@ -100,6 +127,89 @@ func (s *Service) Delete(ctx context.Context) error {
+
+	defer func() {
+		if vm, err := s.VMClient.GetByID(ctx, resourceID); err == nil && vm.VirtualMachineProperties != nil {
+			log.V(4).Info("updating vmss vm state", "state", vm.ProvisioningState)
```
Do we want to be more explicit here and suggest "delete in progress" rather than just the generic "updating vmss vm state"?

CecileRobertMichon commented:
@mboersma what fixed the Flex LB test?

jackfrancis commented on azure/types.go:
```diff
@@ -136,9 +138,24 @@ func (vmss VMSS) InstancesByProviderID() map[string]VMSSVM {
 
 // ProviderID returns the K8s provider ID for the VMSS instance.
 func (vm VMSSVM) ProviderID() string {
+	if vm.IsFlex() {
```
Can we get some UT around this fun?

jackfrancis commented on test/e2e/azure_test.go:
```diff
@@ -556,6 +556,71 @@ var _ = Describe("Workload cluster creation", func() {
+					ClusterctlConfigPath:     clusterctlConfigPath,
+					KubeconfigPath:           bootstrapClusterProxy.GetKubeconfigPath(),
+					InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+					Flavor:                   "external-cloud-provider-machinepool",
```
Would it be better to name this new flavor "external-cloud-provider-vmss-flex" to be clear about the fact that we're testing an Azure scenario (flex) vs a CAPI scenario (machinepool)?

mboersma replied:
That would be more on target for this PR. As I commented elsewhere, I intended this template to test the matrix of Windows|Linux * Uniform|Flexible, but started small. I'll rename it and we can rename it again when its functionality changes.

jackfrancis replied:
Makes sense to me, thanks!

mboersma commented:
> @mboersma what fixed the Flex LB test?

@CecileRobertMichon nothing unfortunately. I commented it out and replaced it with a new test spec that validates providerIDs in an AzureMachinePool to prove that it's either in Flexible or Uniform mode. The LBSpec test was copy-paste from another spec, not a deliberate decision, but it may have uncovered something pointing to cloud-provider-azure. *But* I cannot reproduce it locally if I run just that spec–it succeeded all six times I ran it yesterday.

mboersma commented:
@CecileRobertMichon @jackfrancis I have a working webhook validator that rejects `Flexible` if the Kubernetes version is < v1.26.0. (Validating external-cloud-provider seems problematic, so I skipped that.) I can add that to this or keep it out as a separate PR for followup.

jackfrancis commented:
@mboersma +1 to including it in this PR

CecileRobertMichon commented:
> nothing unfortunately. I commented it out and replaced it with a new test spec that validates providerIDs

that's unfortunate, that LB spec is essentially what validated cloud-provider-azure is working since it exposes a service via LoadBalancer and expects cloud-provider-azure to add the vmss instances to the LB. It's very curious that it passes locally and not on the PR 🤔

CecileRobertMichon commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/retest

mboersma commented:
> It's very curious that it passes locally and not on the PR

It made me wonder if there is somehow test pollution going on, since the LB Spec is reused. But I don't see where that could happen. It's reused several times in the suite.

It failed 100% of the time in CI and succeeded 100% of the time locally, the main differences being I ran only that single test spec locally, and I run with `SKIP_CLEANUP=true`. Let me re-enable the test and try running the full suite locally without that variable.

zmyzheng commented:
> > It's very curious that it passes locally and not on the PR
> 
> It made me wonder if there is somehow test pollution going on, since the LB Spec is reused. But I don't see where that could happen. It's reused several times in the suite.
> 
> It failed 100% of the time in CI and succeeded 100% of the time locally, the main differences being I ran only that single test spec locally, and I run with `SKIP_CLEANUP=true`. Let me re-enable the test and try running the full suite locally without that variable.

Hi @mboersma, is the azure.json file generated by capz? if so, could you check what's the content of the azure.json file? One thing I can think of is you might need to add a new property to that field: "enableVmssFlexNodes": true

Reference: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/tests/k8s-azure/manifest/cluster-api/cloud-config-vmss-default.json#L24

Could you try if this can fix the test failure?

CecileRobertMichon commented:
@zmyzheng why does cloud-provider config require a special property to use flex? Is it a feature flag that is going to be removed in the future?

zmyzheng commented:
> @zmyzheng why does cloud-provider config require a special property to use flex? Is it a feature flag that is going to be removed in the future?

Hi @CecileRobertMichon, this is to prevent querying VmssFlex in the resource group when only VMSS Uniform is used in the cluster to reduce the API calls.  @nilo19 suggests using pure VMSS Uniform in CCM should be the default behavior which aligns with what AKS does today. Let me share some contexts:
CCM needs to remove VMSS from backend pool of loadbalancers (LB) before the LB is deleted. Because of the API difference between Flex and Uniform, CCM batch queries VMSS Flex and VMSS Uniform with two seperate calls and handle the LB backend pool adding/deletion seperately. This means, if we know only VMSS Uniform is used, then there is no need to query the VMSS Flex in the resource group anymore so that some VMSS Flex releated API calls can be saved.

Since CCM is used by both AKS and self-hosted K8s cluster at the same time, @nilo19 suggests my code change in CCM should not affect the behavior or performance of the current AKS pure VMSS Uniform logic. So, he suggests I add a new property "enableVmssFlexNodes" and set it as false by default. In this way, we can assume VMSS Flex is disabled by default and when AKS starts to use the latest CCM, there is no need to handle VMSS Flex related quries. This will make sure the performance of AKS will not be affected when starting using the latest CCM.

This newly added property is documented in this page: https://cloud-provider-azure.sigs.k8s.io/install/configs/

mboersma commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
Seems like Uniform VMSS test is failing...

let's see if it was a flake or an issue with the new property

/retest

mboersma commented:
That uniform test isn't using OOT cloud-provider though, so wouldn't this new property be ignored by the old in-tree code?

zmyzheng commented:
> @zmyzheng I added the `enableVmssFlexNodes: true` to azure.json and the LB spec test finally passed. Thank you!
> 
> What should we do in the (unlikely) case of a cluster that has a Flex VMSS as well as a Uniform VMSS? CAPZ uses one control-plane `azure.json` config and one `azure.json` config for all nodes; we don't have a way currently to customize it for specific nodes.
> 
> That is, if `enableVmssFlexNodes` is enabled on a Uniform VMSS cluster, does it amount to just a de-optimization, or will it cause actual problems?

It will not cause problems. If enableVmssFlexNodes is set to true, the CCM will query the VMSS Flex in the resource group. If it discovers there is no VMSS Flex, then it will not do anything.

CecileRobertMichon commented:
> CAPZ uses one control-plane azure.json config and one azure.json config for all nodes; we don't have a way currently to customize it for specific nodes

The azure.json file is used by the CCM pod, which is shared by all nodes, there isn't a concept of "per node" config

mboersma commented:
/retest

```
Service kube-system/csi-azuredisk-node-win failed
```

I don't think this failure or the previous one were related to the Flex changes, but it is suspicious...

CecileRobertMichon commented:
the last failure is the same flake we just fixed in #2992 

@mboersma could you please squash commits for merge?

mboersma commented:
/test pull-cluster-api-provider-azure-e2e-optional

mboersma commented:
/test pull-cluster-api-provider-azure-test

Linter timeout, yes really.

mboersma commented:
/test pull-cluster-api-provider-azure-e2e
/test pull-clsuter-api-provider-azure-e2e-optional

e2e failed to create an HA cluster and e2e-optional failed on the clusterclass spec. Neither appears to be related to these changes, but that's a worrisome number of flakes. ❄️

mboersma commented:
/test pull-cluster-api-provider-azure-e2e-optional

mboersma commented:
I updated the `azure.json`-related code not to add `enableVmssFlexNodes` if the `EXP_MACHINE_POOLS` feature flag is off (and tested this locally).

mboersma commented:
/test pull-cluster-api-provider-azure-e2e-optional

CecileRobertMichon commented:
/lgtm

CecileRobertMichon reviewed (commented):
/assign @jackfrancis

mboersma commented:
/test pull-cluster-api-provider-azure-e2e

Failed to provision HA cluster.

jackfrancis reviewed (approved):
/lgtm

jackfrancis commented:
/approve

CecileRobertMichon commented:
> Error from server (NotFound): configmaps "kubeadm-config" not found

seems like a race unrelated to this change, will investigate

/retest

mboersma commented:
/retest

I haven't seen this before, yay. I don't think this PR caused it, so retesting.

```go
INFO: "With ipv6 worker node" started at Tue, 10 Jan 2023 20:36:35 UTC on Ginkgo node 6 of 10 and junit test report to file /logs/artifacts/test_e2e_junit.e2e_suite.1.xml
  << Timeline
  [FAILED] Timed out after 1800.000s.
  Expected success, but got an error:
      <*errors.fundamental | 0xc003fa89c0>: {
          msg: "cannot re-use a name that is still in use",
          stack: [0x34dfe66, 0x34dda45, 0x352c418, 0x14ef085, 0x14ee57c, 0x190a67a, 0x190b582, 0x1908d2d, 0x352c09b, 0x351e558, 0x3522130, 0x2f9caf0, 0x35342c8, 0x18e639b, 0x18f9e98, 0x147c741],
      }
      cannot re-use a name that is still in use
  In [It] at: /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/helpers.go:879 @ 01/10/23 20:30:17.977
  Full Stack Trace
    sigs.k8s.io/cluster-api-provider-azure/test/e2e.InstallHelmChart({_, _}, {{0x4223910, 0xc000428190}, {{0xc000d03e30, 0x22}, {0xc000531510, 0x31}, {0xc000533cf3, 0x17}, ...}, ...}, ...)
    	/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/helpers.go:879 +0x5db
    sigs.k8s.io/cluster-api-provider-azure/test/e2e.InstallAzureDiskCSIDriverHelmChart({_, _}, {{0x4223910, 0xc000428190}, {{0xc000d03e30, 0x22}, {0xc000531510, 0x31}, {0xc000533cf3, 0x17}, ...}, ...}, ...)
    	/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/cloud-provider-azure.go:79 +0x1f8
    sigs.k8s.io/cluster-api-provider-azure/test/e2e.EnsureControlPlaneInitialized({_, _}, {{0x4223910, 0xc000428190}, {{0xc000d03e30, 0x22}, {0xc000531510, 0x31}, {0xc000533cf3, 0x17}, ...}, ...}, ...)
    	/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/common.go:268 +0xb90
    sigs.k8s.io/cluster-api/test/framework/clusterctl.ApplyClusterTemplateAndWait({_, _}, {{0x4223910, 0xc000428190}, {{0xc000d03e30, 0x22}, {0xc000531510, 0x31}, {0xc000533cf3, 0x17}, ...}, ...}, ...)
    	/home/prow/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.3.1/framework/clusterctl/clusterctl_helpers.go:334 +0xd30
    sigs.k8s.io/cluster-api-provider-azure/test/e2e.glob..func1.5.1()
    	/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_test.go:303 +0x488
------------------------------
[SynchronizedAfterSuite] PASSED [0.000 seconds]
[SynchronizedAfterSuite] 
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/e2e_suite_test.go:116
```

jackfrancis commented:
@mboersma fwiw I saw that last week (on another PR)

CecileRobertMichon commented:
that shouldn't be happening anymore since I made the helm install idempotent in #2915, I'll look into it

jackfrancis commented:
/retest

jackfrancis commented:
👁️ 🩸
